corehq/util/couch_doc_processor.py

```diff
@@ -0,0 +1,372 @@
+    about an iteration that is in progress. The state will be maintained
+    indefinitely unless it is removed with `discard_state()`.
+    :param chunk_size: Number of documents to yield before updating the
+    iteration checkpoint. In the worst case about this many documents
+    that were previously yielded may be yielded again if the iteration
+    is stopped and later resumed.
+    :param maintain_retry_state: True if the retry state of documents should be maintained
+    after completion of the iterator. If set to True then future runs of the same iterator
+    will continue to iterate over the retry queue until all documents
+    have been successfully processed.
+    """
+
+    def __init__(self, db, doc_types, iteration_key, chunk_size=100, maintain_retry_state=True):
+        if isinstance(doc_types, str):
+            raise TypeError("expected list of strings, got %r" % (doc_types,))
+        self.db = db
+        self.original_doc_types = doc_types = sorted(doc_types)
+        self.iteration_key = iteration_key
+        self.chunk_size = chunk_size
+        iteration_name = "{}/{}".format(iteration_key, " ".join(doc_types))
+        self.iteration_id = hashlib.sha1(iteration_name).hexdigest()
+        self.maintain_retry_state = maintain_retry_state
+        try:
+            self.state = db.get(self.iteration_id)
+        except ResourceNotFound:
+            # new iteration
+            self.state = {
+                "_id": self.iteration_id,
+                "doc_type": "ResumableDocsByTypeIteratorState",
+                "retry": {},
+                "retry_cleared": [],
+
+                # for humans
+                "name": iteration_name,
+                "timestamp": datetime.utcnow().isoformat()
+            }
+            args = {}
+        else:
+            # resume iteration
+            args = self.state.get("offset", {}).copy()
+            if args:
+                assert args.get("startkey"), args
+                doc_type = args["startkey"][0]
+                # skip doc types before offset
+                doc_types = doc_types[doc_types.index(doc_type):]
+            else:
+                # non-retry phase of iteration is complete
+                doc_types = []
+        args.update(
+            view_name='all_docs/by_doc_type',
+            chunk_size=chunk_size,
+            log_handler=ResumableDocsByTypeLogHandler(self),
+            include_docs=True,
+            reduce=False,
+        )
+        self.view_args = args
+        self.doc_types = doc_types
+
+    def __iter__(self):
+        args = self.view_args
+        for doc_type in self.doc_types:
+            if args.get("startkey", [None])[0] != doc_type:
+                args.pop("startkey_docid", None)
+                args["startkey"] = [doc_type]
+            args["endkey"] = [doc_type, {}]
+            for result in paginate_view(self.db, **args):
+                yield result['doc']
+
+        retried = {}
```

Comment:
Proposal for maintaining retry state across stop/resume:
```py
retried = {}
try:
    while self.state["retry"] != retried:
        ...
except:
    exc_type, exc_value, exc_tb = sys.exc_info()
    for doc_id, retries in list(self.state["retry"].iteritems()):
        if retries == retried.get(doc_id):
            # clear doc that has successfully been retried
            self.state["retry"].pop(doc_id)
    self._save_state()
    raise exc_type, exc_value, exc_tb
```
This might not work: there might be an edge case here where a retried document will be dropped from the retry state if the loop is exited due to too many retries. I think `retries == retried.get(doc_id)` in that case. Not sure what to do about that one. Maybe don't remove the item from the retry state if it was the `doc_id` being processed when the exception was raised? That seems like it could get messy...

The docstring for `retry` will need to be updated, and unfortunately retry behavior on stop/resume will not be quite as simple to explain. I think we can just change "_every retry document will be yielded each time the iterator is stopped and resumed_" to "_every retry document may be yielded each time the iterator is stopped and resumed_".